livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java

```diff
@@ -284,12 +275,25 @@ public InterpreterResult interpret(String code,
       }
     } finally {
       if (paragraphId != null) {
-        paragraphId2StmtIdMap.remove(paragraphId);
         paragraphId2StmtProgressMap.remove(paragraphId);
       }
     }
   }
 
+  private void cancel(int id, String paragraphId) {
+    if (livyVersion.isCancelSupported()) {
+      try {
+        LOGGER.info("Cancelling statement " + id);
+        cancelStatement(id);
+        paragraphsToCancel.remove(paragraphId);
```

Comment:
I have put the removal in finally block so that paragraph is removed from the set even if cancellation is not supported in the livy version.